core/src/main/java/org/apache/druid/data/input/RetryingInputEntity.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.data.input;
+
+import com.google.common.base.Predicate;
+import org.apache.druid.data.input.impl.RetryingInputStream;
+import org.apache.druid.data.input.impl.prefetch.ObjectOpenFunction;
+import org.apache.druid.java.util.common.RetryUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+public interface RetryingInputEntity extends InputEntity
+{
+  @Override
+  default InputStream open() throws IOException
+  {
+    return new RetryingInputStream<>(
+        this,
+        new RetryingInputEntityOpenFunction(),
+        getRetryCondition(),
+        RetryUtils.DEFAULT_MAX_TRIES
+    );
+  }
+
+  /**
+   * Directly opens an {@link InputStream} on the input entity.
+   */
+  default InputStream readFromStart() throws IOException
+  {
+    return readFrom(0);
+  }
+
+  /**
+   * Directly opens an {@link InputStream} starting at the given offset on the input entity.
+   * This is the basic way to read the given entity.
+   */
+  InputStream readFrom(long offset) throws IOException;
```

Comment:
Would there ever be cases where the offset type for an InputEntity might be something other than a `long`?

If not, can you add a comment about that?

Reply:
Hmm good question. It's hard for me to imagine an offset of other types since `InputEntity` abstracts an byte-representable entity. I added javadoc on it.

Reply:
Cool, thanks